library/core/src/time.rs

```diff
@@ -1698,3 +1728,213 @@ impl Duration {
         )
     }
 }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+pub mod duration_verify {
+    use crate::kani;
+    use super::*;
+
+    impl kani::Arbitrary for Duration {
+        fn any() -> Duration {
+            let secs = kani::any::<u64>();
+            let nanos = kani::any::<u32>();
+            Duration::new(secs, nanos)
+        }
+    }
+
+    fn safe_duration() -> Duration {
+        let secs = kani::any::<u64>();
+        let nanos = kani::any::<u32>();
+        kani::assume(nanos < NANOS_PER_SEC || secs.checked_add((nanos / NANOS_PER_SEC) as u64).is_some());
```

Comment:
Why do we need this assumption in the harness?

Reply:
This assumption constrains the inputs to `Duration::new()`, so that the call never `panics`. The constraints were taken from the implementation of `Duration::new()`:

```rust
pub const fn new(secs: u64, nanos: u32) -> Duration {
        if nanos < NANOS_PER_SEC {
            // SAFETY: nanos < NANOS_PER_SEC, therefore nanos is within the valid range
            Duration { secs, nanos: unsafe { Nanoseconds(nanos) } }
        } else {
            let secs = secs
                .checked_add((nanos / NANOS_PER_SEC) as u64)
                .expect("overflow in Duration::new");
            let nanos = nanos % NANOS_PER_SEC;
            // SAFETY: nanos % NANOS_PER_SEC < NANOS_PER_SEC, therefore nanos is within the valid range
            Duration { secs, nanos: unsafe { Nanoseconds(nanos) } }
        }
    }
```

You can see that the only time where `Duration::new()` panics is when `secs.checked_add((nanos / NANOS_PER_SEC) as u64)` is `None`. 

We added this non-panicking harness so that we could have panicking and non-panicking contracts.

Reply:
Why shouldn't we capture this condition as part of the requires clause for `Duration::new`?

Reply:
Micheal left comments above stating that `safe` functions shouldn't have `requires` clauses. So, we moved the assumption to the guarded function `safe_duration` which guarantees that `Duration::new()` won't panic.